### PR TITLE
Unit of work

### PR DESCRIPTION
**Pull request by Vaub**
Simple unit of work for testing and servlets.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,6 +89,33 @@ MyClass instance = ServiceLocator.locate().resolve(MyClass.class);
...
+unitOfWork.begin();
+
+try {
+	// never use simple double for currency manipulations, this is an example
+	unitOfWork.withTransaction(transaction -> withdraw("user1", 5.0));
+	unitOfWork.withTransaction(transaction -> withdraw("user2", 5.0));
+	unitOfWork.withTransaction(transaction -> transfer("user3", 10.0));
+
+	unitOfWork.commit();
+} catch (CriticalMoneyException e) {
```
Comment (ALWAL12) [inline README.md]: Even though the example is pretty good, I think I would not use a money-related example and simplify the example. It is not complex, but I feel I need to argue with the double topic 🤓

File: src/main/java/ca/nexapp/core/persistence/unitofwork/exceptions/UnitOfWorkException.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package ca.nexapp.core.persistence.unitofwork.exceptions;
+
+public class UnitOfWorkException extends RuntimeException {
```
Comment (ALWAL12) [inline src/main/java/ca/nexapp/core/persistence/unitofwork/exceptions/UnitOfWorkException.java]: Your exceptions must generate a serializableUUID

File: src/test/java/ca/nexapp/core/persistence/unitofwork/UnitOfWorkSuiteITest.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.function.Function;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import ca.nexapp.core.persistence.unitofwork.exceptions.AlreadyInTransactionException;
+import ca.nexapp.core.persistence.unitofwork.exceptions.NotInTransactionException;
+
+public abstract class UnitOfWorkSuiteITest<T extends AutoCloseable> {
```
Comment (ALWAL12) [inline src/test/java/ca/nexapp/core/persistence/unitofwork/UnitOfWorkSuiteITest.java]: It is not a `Suite`. I would simply call it `UnitOfWorkTest`. The implementation will indicate if it is a unit or integration test.

File: src/test/java/ca/nexapp/core/persistence/unitofwork/UnitOfWorkSuiteITest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    private static final int A_KEY = 5;
+    private static final String A_VALUE = "Hello World!";
+
+    protected UnitOfWork<T> unitOfWork;
```
Comment (ALWAL12) [inline src/test/java/ca/nexapp/core/persistence/unitofwork/UnitOfWorkSuiteITest.java]: why not `private`?

File: src/test/java/ca/nexapp/core/persistence/unitofwork/rest/TransactionFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+@RunWith(MockitoJUnitRunner.class)
+public class TransactionFilterTest {
+
+    @Mock
+    private UnitOfWork<Connection> unitOfWork;
+
+    private ServletRequest servletRequest;
+    private ServletResponse servletResponse;
+    private FilterChain filterChain;
```
Comment (ALWAL12) [inline src/test/java/ca/nexapp/core/persistence/unitofwork/rest/TransactionFilterTest.java]: Why do you use `@Mock` for the first item, but not for the others here?

File: src/test/java/ca/nexapp/core/persistence/unitofwork/rest/TransactionFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        servletResponse = mock(ServletResponse.class);
+        filterChain = mock(FilterChain.class);
+
+        transactionFilter = new TransactionFilter<>(unitOfWork);
+    }
+
+    @Test
+    public void shouldBeginATransaction() throws IOException, ServletException {
+        transactionFilter.doFilter(servletRequest, servletResponse, filterChain);
+        verify(unitOfWork).begin();
```
Comment (ALWAL12) [inline src/test/java/ca/nexapp/core/persistence/unitofwork/rest/TransactionFilterTest.java]: All tests must have a blank line between the _given-when-then_ block. 🤓

File: src/test/java/ca/nexapp/core/persistence/unitofwork/rest/TransactionFilterTest.java (hunk 5 above)
Comment (Vaub) [reply src/test/java/ca/nexapp/core/persistence/unitofwork/rest/TransactionFilterTest.java]: limitation of the `mock(...)` methods and Java type-system in general. Easiest method to mock a templated class

File: src/main/java/ca/nexapp/core/persistence/unitofwork/UnitOfWork.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package ca.nexapp.core.persistence.unitofwork;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import ca.nexapp.core.persistence.unitofwork.exceptions.AlreadyInTransactionException;
+import ca.nexapp.core.persistence.unitofwork.exceptions.NotInTransactionException;
```
Comment (ALWAL12) [inline src/main/java/ca/nexapp/core/persistence/unitofwork/UnitOfWork.java]: The import should not present in this interface's file, and I just noticed you created a pakcage `exception` which is useless for now.

**Pull request closed without merging**